File: cpp/backtracking/permutation.cpp
#include <vector>

using std::vector;

// Time: O(n^2 * n!)
vector<vector<int>> permutationsRecursive(vector<int>& nums) {
    return helper(0, nums);
}

vector<vector<int>> helper(int i, vector<int>& nums) {
    if (i == nums.size()) {
        return {{}};
    }
    vector<vector<int>> resPerms;
    vector<vector<int>> perms = helper(i + 1, nums);
    for (vector<int> p : perms) {
        for (int j = 0; j < p.size() + 1; j++) {
            vector<int> pCopy(p);
            pCopy.insert(pCopy.begin() + j, nums[i]);
            resPerms.push_back(pCopy);
        }
    }
    return resPerms;
}


// Time: O(n^2 * n!)
vector<vector<int>> permutationsIterative(vector<int>& nums) {
    vector<vector<int>> perms = {{}};
        
    for (int n : nums) {
        vector<vector<int>> nextPerms;
        for (vector<int> p : perms) {
            for (int i = 0; i < p.size() + 1; i++) {
                vector<int> pCopy(p);
                pCopy.insert(pCopy.begin() + i, n);
                nextPerms.push_back(pCopy);
            }
        }
        perms = nextPerms;
    }
    return perms;
}

File: swift/backtracking/permutation.swift
// Time: O(n^2 * n!)
func permutationsRecursive(_ nums: [Int]) -> [[Int]] {
  return helper(0, nums)
}

func helper(_ i: Int, _ nums: [Int]) -> [[Int]] {
  if i == nums.count {
    return [[]]
  }

  var resPerms: [[Int]] = []
  let perms = helper(i + 1, nums)
  for p in perms {
    for j in 0..<p.count + 1 {
      var pCopy = p
      pCopy.insert(nums[i], at: j)
      resPerms.append(pCopy)
    }
  }
  return resPerms
}


// Time: O(n^2 * n!)
func permutationsIterative(_ nums: [Int]) -> [[Int]] {
  var perms: [[Int]] = [[]]

  for n in nums {
    var nextPerms: [[Int]] = []
    for p in perms {
      for i in 0..<p.count + 1 {
        var pCopy = p
        pCopy.insert(n, at: i)
        nextPerms.append(pCopy)
      }
    }
    perms = nextPerms
  }
  return perms
}